### ISSUE TITLE
Freeforms issue

### ISSUE THREAD

**Issue by jtraczms** [microsoft/lightgbm-transform#22]
## Description
When trying to leverage freeforms in learning to rank scenario it seems they are working only when the features used in them are also defined in parser config file as linear Transforms (last example from the attached notebook). Using freeforms with features not defined as standalone seems like the ranker does not use them at all.

## Reproducible example
[ff_example.zip](https://github.com/microsoft/lightgbm-transform/files/9415436/ff_example.zip)

## Environment info

lightgbm-transform version 3.3.1

Command(s) you used to install lightgbm-transform

Using docker image with 
`RUN pip install --upgrade pip setuptools wheel && \
    pip install 'cmake==3.21.0' && \
    pip install lightgbm==3.3.1 --install-option=--mpi &&\
    pip install lightgbm-transform==3.3.1`

**Comment by ltxtech1999** [comment]
Hi,@jtraczms thanks for asking and providing a repro case. I‘m looking into the issue, and have some questions:

- Could you please clarify more about "Using freeforms with features not defined as standalone seems like the ranker does not use them at all."? The last example goes well and logs record there are 6 feature used in the scenario, matching 6 features in "baseline_ff".
- From your example notebook, you say f5 is the only relevant feature that will affect the final result. Could you please tell me why? Seems that if we remove other features(eg: f1), the result will also change. This means f5 is not the only one that can affect the training process.

**Comment by jtraczms** [comment]
Hi @ltxtech1999, to answer your questions:
- Last notebook example is just for showing that using freeforms on features works when the features are also used as linear transformation. Second to last example is the issue I'm reporting: when using f5 (header "No f5 in input") only as a freeform model cannot learn anything meaningful (while I would expect it to learn just as good as in baseline, where f5 is provided in input section as Linear transform).
- I've generated the labels based on the f5 argmax ` y[i*40 + np.argmax(X[i*40:(i+1)*40, 4])] = 1`, so the rest of the features is noise. Most likely it can affect the results due to memorization (we are validating on the training set after all in this example)

**Comment by ltxtech1999** [comment]
Hi, @jtraczms we could repro the issue when changing "Input 6" transform type from "Linear" to "FreeForm2".

This is because FreeForm2 type is designed to accept unsigned int (See https://github.com/microsoft/lightgbm-transform/blob/main/src/TransformProcessor/src/TransformProcessor.cpp#L198 for details) while your dummy test data "f5" column has decimal places.

FreeForm2 parser is designed based on common use cases over MS products, where features are usually aggregated num like click/session counts, so they are of unsigned int type for both computation efficiency and actual needs. We will help add this limitation description to docs.

If you really need decimal type support, feel free to send pull request to improve current FreeForm2Parser or create a new parser. See ["LightGBM Customized Parser Tutorial"](https://github.com/microsoft/lightgbm-transform/blob/main/docs/Customized-Parser-Tutorial.rst) for parser customization.

**Comment by jtraczms** [comment]
Thanks for the clarification on the input types!

I do have another question regarding the freefroms; If I use the freeforms division on the f5 (when inputs are ints) I can see performance degradation - is the uint limitation only with respect to inputs or does it also concern the output of the transformation? 

If it's not clear I can prepare an example

**Comment by jtraczms** [comment]
The issue I was experiencing was due to how I prepared the freeform transform - it turns out for the division to work properly I had to use float value in denominator. For future reference; this did not work:
`Transform=FreeForm2\nLine1=(/ f5 1000)` for f5 in range (0, 1000) 
While this did:
`Transform=FreeForm2\nLine1=(/ f5 1000.0)`
Thanks for the help!
